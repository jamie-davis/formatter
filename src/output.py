from ObjectAnalysis.column_classifier import classify_columns
from ObjectAnalysis.field_extractor import FieldExtractor


def map_values(values, columns):
    col_index = 0
    num_columns = len(columns)
    value = iter(values)
    line = ''
    for col in columns:
        col_index += 1

        col_value = str(next(value))
        if not col.left_align:
            col_value = col_value.rjust(col.required_width)

        if (col_index < num_columns):
            col_value = col_value.ljust(col.required_width + 1)

        line += col_value

    return line + "\n"


class Output:
    """ Class for formatting text data. Call the methods of the class to fill the report buffer and then access the
    formatted data using the report property."""
    def __init__(self):
        self.report = ''

    def format_table(self, items):
        """Format a list of objects as a report"""
        fields = FieldExtractor.extract(items)
        columns = classify_columns(fields)

        if any([col.heading for col in columns if col.heading is not None]):
            self.report += map_values([col.heading if col.heading is not None else '' for col in columns], columns)
            self.report += map_values([('-' * col.required_width) for col in columns], columns)

        for item in fields.rows:
            values = [item[col.heading] if col.heading in item.keys() else '' for col in columns]
            self.report += map_values(values, columns)
