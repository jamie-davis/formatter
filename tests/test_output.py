import pytest

from output import Output
from approvaltests import verify


def test_report():
    output = Output()
    assert output.report == ""


def test_format_table():
    """Test the simplest table of objects"""

    class TestItem:
        def __init__(self, name, age, num_pets):
            self.name = name
            self.age = age
            self.num_pets = num_pets

    # Arrange
    items = [TestItem("Ann", 21, 1), TestItem("Bob", 45, 2), TestItem("Cath", 52, 4)]
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_not_collection():
    # Arrange
    items = "not an array"
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_from_object():
    # Arrange
    class TestItem1:
        def __init__(self, name, age, num_pets):
            self.name = name
            self.age = age
            self.num_pets = num_pets

    items = TestItem1("Ann", 21, 1)
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_with_sparse_columns():
    """Test table where rows have different heading names"""

    class TestItem1:
        def __init__(self, name, age, num_pets):
            self.name = name
            self.age = age
            self.num_pets = num_pets

    class TestItem2:
        def __init__(self, name, age, num_computers):
            self.name = name
            self.age = age
            self.num_computers = num_computers

    # Arrange
    items = [TestItem1("Ann", 21, 1), TestItem2("Bob", 45, 2), TestItem1("Cath", 52, 4)]
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_with_list_of_primitives():
    # Arrange
    items = [1, 2000, 3, 4]
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_with_list_of_mixed_primitives():
    # Arrange
    items = [1, 2, 'banana', 300, 4]
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_with_list_of_primitives_and_objects():
    class TestItem:
        def __init__(self, name, age, num_pets):
            self.name = name
            self.age = age
            self.num_pets = num_pets

    # Arrange
    items = [TestItem("Ann", 21, 1), 4500, 'banana', TestItem("Cath", 52, 4)]
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)


def test_format_table_with_list_of_dict():
    # Arrange
    items = [{
        'one': 1,
        'two': 2,
    }]
    output = Output()

    # Act
    output.format_table(items)

    # Assert
    verify(output.report)
