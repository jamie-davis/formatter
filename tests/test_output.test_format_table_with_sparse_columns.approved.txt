name age num_pets num_computers
---- --- -------- -------------
Ann   21        1              
Bob   45                      2
Cath  52        4              
